"InvalidOperationException: Collection was modified" when executing ObjectPool.Remove

So I'm trying to remove a list of submenus to replace them with new ones and I'm getting this error. Sometimes also happens after adding submenus.

@justalemon 

[19:17:35] [DEBUG] Loading API from .\ScriptHookVDotNet2.dll ...
[19:17:35] [DEBUG] Loading API from .\ScriptHookVDotNet3.dll ...
[19:17:35] [DEBUG] Loading scripts from C:\Program Files (x86)\Steam\steamapps\common\Grand Theft Auto V\scripts ...
[19:17:35] [DEBUG] Loading assembly LemonUI.SHVDN3.dll ...
[19:17:35] [INFO] Found 0 script(s) in LemonUI.SHVDN3.dll.
[19:17:35] [DEBUG] Loading assembly Meta.dll ...
[19:17:35] [INFO] Found 1 script(s) in Meta.dll resolved to API 3.4.0.
[19:17:35] [DEBUG] Instantiating script Main ...
[19:17:38] [INFO] Started script Main.
[19:17:41] [ERROR] Caught fatal unhandled exception:
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at LemonUI.ObjectPool.Process()
   at Main.OnTick(Object sender, EventArgs e)
   at System.EventHandler.Invoke(Object sender, EventArgs e)
   at SHVDN.Script.MainLoop()
[19:17:41] [ERROR] The exception was thrown while executing the script Main.
[19:17:41] [WARNING] Aborted script Main.

The exception happens when adding an item to the ObjectPool in a different thread, such as when adding items to hooked events, like the Opening event handler por example.

The way I see it, you should implement a manual concurrency with locks, or replace the objects data type by a thread safe structure, e.g. ConcurrentDictionary.

`private readonly List<IProcessable> objects = new List<IProcessable>();`

to

`private readonly ConcurrentDictionary<IProcessable> objects = new();`

https://docs.microsoft.com/pt-br/dotnet/standard/collections/thread-safe/

--- a/LemonUI/ObjectPool.cs
+++ b/LemonUI/ObjectPool.cs
@@ -13,7 +13,7 @@ using GTA.Native;
 using GTA.UI;
 #endif
 using System;
-using System.Collections.Generic;
+using System.Collections.Concurrent;
 using System.Drawing;
 
 namespace LemonUI
@@ -106,7 +106,7 @@ namespace LemonUI
         /// <summary>
         /// The list of processable objects.
         /// </summary>
-        private readonly List<IProcessable> objects = new List<IProcessable>();
+        private readonly ConcurrentDictionary<int, IProcessable> objects = new ConcurrentDictionary<int, IProcessable>();
 
         #endregion
 
@@ -120,7 +120,7 @@ namespace LemonUI
             get
             {
                 // Iterate over the objects
-                foreach (IProcessable obj in objects)
+                foreach (IProcessable obj in objects.Values)
                 {
                     // If is visible return true
                     if (obj.Visible)
@@ -221,12 +221,13 @@ namespace LemonUI
                 throw new ArgumentNullException(nameof(obj));
             }
 
+            int key = obj.GetHashCode();
             // Otherwise, add it to the general pool
-            if (objects.Contains(obj))
+            if (objects.ContainsKey(key))
             {
                 throw new InvalidOperationException("The object is already part of this pool.");
             }
-            objects.Add(obj);
+            objects.TryAdd(key, obj);
         }
         /// <summary>
         /// Removes the object from the pool.
@@ -234,7 +235,7 @@ namespace LemonUI
         /// <param name="obj">The object to remove.</param>
         public void Remove(IProcessable obj)
         {
-            objects.Remove(obj);
+            objects.TryRemove(obj.GetHashCode(), out _);
         }
         /// <summary>
         /// Performs the specified action on each element that matches T.
@@ -243,7 +244,7 @@ namespace LemonUI
         /// <param name="action">The action delegate to perform on each T.</param>
         public void ForEach<T>(Action<T> action)
         {
-            foreach (IProcessable obj in objects)
+            foreach (IProcessable obj in objects.Values)
             {
                 if (obj is T conv)
                 {
@@ -257,7 +258,7 @@ namespace LemonUI
         public void RefreshAll()
         {
             // Iterate over the objects and recalculate those possible
-            foreach (IProcessable obj in objects)
+            foreach (IProcessable obj in objects.Values)
             {
                 if (obj is IRecalculable recal)
                 {
@@ -270,7 +271,7 @@ namespace LemonUI
         /// </summary>
         public void HideAll()
         {
-            foreach (IProcessable obj in objects)
+            foreach (IProcessable obj in objects.Values)
             {
                 obj.Visible = false;
             }
@@ -285,7 +286,7 @@ namespace LemonUI
             DetectResolutionChanges();
             DetectSafezoneChanges();
             // And process the objects in the pool
-            foreach (IProcessable obj in objects)
+            foreach (IProcessable obj in objects.Values)
             {
                 obj.Process();
             }

